src/ui/processPicker.ts

```diff
@@ -0,0 +1,252 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+'use strict';
+
+import * as nls from 'vscode-nls';
+import * as vscode from 'vscode';
+import { basename } from 'path';
+import { getProcesses } from './processTree';
+import { execSync } from 'child_process';
+import { Contributions } from '../common/contributionUtils';
+import {
+  nodeAttachConfigDefaults,
+  INodeAttachConfiguration,
+  ResolvingNodeAttachConfiguration,
+} from '../configuration';
+import { guessWorkingDirectory } from '../nodeDebugConfigurationProvider';
+import { mapValues } from '../common/objUtils';
+
+const INSPECTOR_PORT_DEFAULT = 9229;
+
+const localize = nls.loadMessageBundle();
+
+interface ProcessItem extends vscode.QuickPickItem {
+  pidOrPort: string; // picker result
+  sortKey: number;
+}
+
+/**
+ * end user action for picking a process and attaching debugger to it
+ */
+export async function attachProcess() {
+  let config: INodeAttachConfiguration = {
+    ...nodeAttachConfigDefaults,
+    name: 'process',
+    processId: `\${command:${Contributions.PickProcessCommand}}`,
+  };
+
+  if (!(await resolveProcessId(config))) {
+    return;
+  }
+
+  const cwd = guessWorkingDirectory(config);
+  const assignWorkspaceFolder = (obj: any) =>
```

Comment:
I wasn't super happy with this... in our defaults we have some thing, like sourcemap overrides, that map based on the workspaceFolder, and if you attach to a process in a multiroot workspace VS Code will yell at us saying it can't find a valid folder. This goes through and replaces instances of the `${workspaceFolder}` with the one we guessed. (Guessing logic is the same as what the config resolution already uses and used in v1/2 of the debug adapter)

Reply:
`${workspaceFolder}` in a launch config should actually be handled by vscode, not the DA.

Reply:
Yea, but in a multi-root workspace where you use the process picker to choose a process to attach to (i.e. not running out of a launch.json), VS Code was failing to resolve the workspace folder. Do you know if there's a better approach here?

Reply:
Where is `${workspaceFolder}` coming from in this case?

Reply:
It comes from the configuration defaults. e.g. sourceMapOverrides depends on it, to resolve `webpack:///foo/bar.js` to an absolute path.

Reply:
Oh ok. I guess I don't know what to do about that. We can leave it for now.